### PR TITLE
Network activity toggle

### PR DESCRIPTION
**Pull request by luke-jr**
Yet another rebase of #2412 / #5314

I've made the RPC interface more consistent (booleans only) and removed the `getinfo` addition.

Also removed the problematically-licensed icon and replaced it with an equivalent using our new style, from Typfonts.

I'd like to replace the PNG with its SVG equivalent, but I don't know if this is safe right now...?

### CODE REVIEW

**Comment by paveljanik** [comment]
Concept ACK

While trying to test this, I have got:
```
Assertion failed: (nMaxInbound > 0), function AcceptConnection, file net.cpp, line 973.
```
almost immediately I pressed the network toggle icon.

When the network is toggled off, the icon is hidden and I can't turn the network back on without trying to click in the empty space...

**Comment by paveljanik** [comment]
There is a commit for RPC `togglenetwork` (f83a1e2bd83625997916963fe8b1942e504070ab), but it is renamed later, so commits do not match.

The RPC test is still named `togglenetwork` (https://github.com/bitcoin/bitcoin/pull/8996/files#diff-b89234789a9a9ee29cb0758f69a73133R87).

**Comment by paveljanik** [comment]
During startup, when there is 1 outgoing connection only:

```
2016-10-23 16:44:42 SetNetworkActive: false
2016-10-23 16:44:42 disconnecting peer=1
2016-10-23 16:44:42 socket select error Bad file descriptor (9)
2016-10-23 16:44:42 disconnecting peer=2
nMaxConnections = 8
nMaxFeeler = 1
nMaxOutbound = 8
nMaxInbound = -1
Assertion failed: (nMaxInbound > 0), function AcceptConnection, file net.cpp, line 979.
```
@EthanHeilman Do you have an idea?

**Comment by luke-jr** [comment]
> ```Assertion failed: (nMaxInbound > 0), function AcceptConnection, file net.cpp, line 973.```

I don't believe this can be related to this PR...

> When the network is toggled off, the icon is hidden and I can't turn the network back on without trying to click in the empty space...

Sounds like a build system problem - it didn't update the resources with the new icon?

> There is a commit for RPC togglenetwork (f83a1e2), but it is renamed later, so commits do not match.

I don't see this as a problem.

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -1211,3 +1231,18 @@ void UnitDisplayStatusBarControl::onMenuSelection(QAction* action)
...
+
+void NetworkToggleStatusBarControl::mousePressEvent(QMouseEvent *event)
+{
+    if (clientModel) {
+        clientModel->setNetworkActive(!clientModel->getNetworkActive());
+    }
+}
+
+/** Lets the control know about the Client Model */
+void NetworkToggleStatusBarControl::setClientModel(ClientModel *clientModel)
```
Comment (paveljanik) [inline src/qt/bitcoingui.cpp]: `_clientModel` please to prevent shadowing.

**Comment by paveljanik** [comment]
The missing icon issue was local one, yes (solved by clean build). Sorry for confusion.

The icon should probably contain some visual reference to the network (compare with the no-HD icon). I can't create screenshot because it aborts almost immediately here.

**Comment by paveljanik** [comment]
It works as designed. Icon in the UI flips on RPC calls, can be clicked, nice tooltips, it works.

When the network is disabled, it looks like this (with non-HD wallet):

screen shot 2016-10-24 at 18 58 52

I still think that it should be done in 2 commits instead of this mess:

- RPC: New method setnetworkactive
- GUI: UI part of the network enable/disable

**Comment by EthanHeilman** [comment]
@paveljanik At the risk of stating the obvious if nMaxConnections is set to 8 then nMaxInbound will be set to -1 triggering the assert.

```c++
int nMaxInbound = nMaxConnections - (nMaxOutbound + nMaxFeeler);
assert(nMaxInbound > 0);
```

The value nMaxConnections is typically set to 125, I don't understand how it is getting set to 8, perhaps the system is running low on available sockets?

**Comment by sipa** [comment]
@EthanHeilman nMaxConnections is user configurable. I typically set it to 3 on low-memory devices.

**Comment by luke-jr** [comment]
I think the assert failure with a low maxconnections is an issue unrelated to this PR.

**Comment by sipa** [comment]
@luke-jr @EthanHeilman Indeed, let's move to #9007.

**Comment by paveljanik** [comment]
I think this is very useful function for some users. Please review.

**Comment by paveljanik** [comment]
@jonasschnelli Jonas, can you please have a look?

**Comment by jonasschnelli** [comment]
The reason why I'm waiting for this are the conceptual NACK's on the previous attempt to get this in: https://github.com/bitcoin/bitcoin/pull/5314

Some devs where claiming that we need a better solution then just disabling the network connection. Ideally, this mode should connect to the p2p network in case you have uncommitted wtxs.

I'd like to get more Concept ACKs from others.
I always likes this solution as a first step:

Concept ACK.

**Comment by paveljanik** [comment]
The previous NACK (there was one ;-) was about `togglenetwork` interface. This was removed.

Wallet showing wrong info without network is the current state anyway. We could slide the "gray overlay" up as we do during IBD to make it clear!

**Comment by paveljanik** [comment]
Some more thoughts about this: it depends how you look at this. So far, my view was as simple as follows. I do not care at all about the displayed stuff in the GUI, I just want GUI to immediately stop all communication. Be it if I plan to connect to the untrusted network, or slow network at parents', or limited usage network where every eight connections to the outside mean that no one else being able to use the network... I was solving this by suspending the GUI process (I do not want to stop it and start again when I need it). This brings new and elegant solution!

The "better solution" mentioned above probably wants to solve different problem(s) though.

**Comment by jonasschnelli** [comment]
Tested ACK 19f46f177ec5e1913f9be5b257dad95bc7a57c38.

Would be nice if we would at least add a tooltip "Press to enable/disable network activity" over the connection statusbar icon. From the GUI perspective, it's kind of a hidden feature right now.

File: src/net.cpp [hunk 2]
```diff
@@ -2024,8 +2033,30 @@ void Discover(boost::thread_group& threadGroup)
...
 }
 
+void CConnman::SetNetworkActive(bool active)
+{
+    if (fDebug) {
+        LogPrint("net", "SetNetworkActive: %s\n", active);
+    }
+
+    if (!active) {
+        fNetworkActive = false;
```
Comment (sipa) [inline src/net.cpp]: No locking?

Comment (jonasschnelli) [inline src/net.cpp]: Oh! I though this was a `std::atomic<bool>`, but it's a plain bool.

**Comment by jonasschnelli** [comment]
Sorry. This merge was a little bit pro active.

**Pull request merged**